File: Maintainers/agentsquared-skills-helper/scripts/validate_runtime_contract.py
#!/usr/bin/env python3

from __future__ import annotations

import sys
from pathlib import Path


ROOT = Path(__file__).resolve().parents[3]


FORBIDDEN_SNIPPETS = {
    "removed onboarding guide endpoint": "`https://api.agentsquared.net/api/onboard`",
    "removed relay public soul endpoint": "/api/relay/friends/agents/{agentName@humanName}/public-soul",
    "removed relay public memory endpoint": "/api/relay/friends/agents/{agentName@humanName}/public-memory",
    "removed onboarding displayName field": '"displayName"',
}


REQUIRED_SUBSTRINGS = {
    ROOT / "Base" / "p2p-session-handoff" / "SKILL.md": [
        "targetTransport",
        "agentCard.preferredTransport",
        "connectTicket",
        "POST /api/relay/connect-tickets/introspect",
    ],
    ROOT / "Base" / "p2p-session-handoff" / "package.json": [
        '"name": "agentsquared-p2p-session-handoff"',
        '"self-test": "node ./scripts/self_test.mjs"',
    ],
    ROOT / "Base" / "gateway" / "SKILL.md": [
        "shared long-lived inbound listener/router",
        "trusted peer-session reuse",
        "default to `friend-im`",
        "local-only control endpoint",
        "serve_agent_router.mjs",
    ],
    ROOT / "Base" / "gateway" / "scripts" / "serve_gateway.mjs": [
        "/health",
        "/inbound/next",
        "/inbound/respond",
        "/inbound/reject",
        "/connect",
        "runtimeState",
    ],
    ROOT / "Base" / "gateway" / "scripts" / "serve_agent_router.mjs": [
        "max-active-mailboxes",
        "gatewayNextInbound",
        "createMailboxScheduler",
        "buildSkillResult",
    ],
    ROOT / "Identity" / "agent-onboarding" / "SKILL.md": [
        "../../Base/gateway/SKILL.md",
        "gateway status",
        "router status",
    ],
    ROOT / "Identity" / "agent-onboarding" / "references" / "onboarding-contract.md": [
        "Base/gateway/scripts/serve_gateway.mjs",
        "Base/gateway/scripts/serve_agent_router.mjs",
        "only send later relay MCP requests after confirming the local listener is still active",
    ],
    ROOT / "Friends" / "friend-im" / "SKILL.md": [
        "../../Base/p2p-session-handoff/SKILL.md",
        "../../Base/gateway/scripts/serve_gateway.mjs",
        "private peer payload",
    ],
    ROOT / "Friends" / "agent-mutual-learning" / "SKILL.md": [
        "../../Base/p2p-session-handoff/SKILL.md",
        "../../Base/gateway/scripts/serve_gateway.mjs",
        "private session",
    ],
    ROOT / "Friends" / "friend-im" / "scripts" / "send_friend_im.mjs": [
        "skillHint: 'friend-im'",
        "gatewayConnect",
    ],
    ROOT / "Friends" / "agent-mutual-learning" / "scripts" / "start_mutual_learning.mjs": [
        "skillHint: 'agent-mutual-learning'",
        "gatewayConnect",
    ],
    ROOT / "Base" / "p2p-session-handoff" / "scripts" / "lib" / "peer_session.mjs": [
        "relayConnectTicket",
        "createConnectTicket",
        "introspectConnectTicket",
        "reusedSession",
        "peerSessionId",
    ],
}


def iter_skill_text_files(root: Path) -> list[Path]:
    files: list[Path] = []
    for path in root.rglob("*"):
        if not path.is_file():
            continue
        if ".git" in path.parts:
            continue
        if path.name == "validate_runtime_contract.py":
            continue
        if path.suffix not in {".md", ".yaml", ".json", ".py"}:
            continue
        files.append(path)
    return files


def main() -> int:
    errors: list[str] = []
    files = iter_skill_text_files(ROOT)

    for path in files:
        text = path.read_text(encoding="utf-8")
        for label, needle in FORBIDDEN_SNIPPETS.items():
            if needle in text:
                errors.append(f"{path}: still contains {label}: {needle}")

    for path, needles in REQUIRED_SUBSTRINGS.items():
        text = path.read_text(encoding="utf-8")
        for needle in needles:
            if needle not in text:
                errors.append(f"{path}: missing required text: {needle}")

    if errors:
        print("AgentSquared Skills runtime-contract validation failed:")
        for err in errors:
            print(f"- {err}")
        return 1

    print("AgentSquared Skills runtime-contract validation passed.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: Maintainers/agentsquared-skills-helper/scripts/validate_runtime_contract.mjs
#!/usr/bin/env node

import fs from 'node:fs'
import path from 'node:path'
import { fileURLToPath } from 'node:url'

const __filename = fileURLToPath(import.meta.url)
const __dirname = path.dirname(__filename)
const ROOT = path.resolve(__dirname, '..', '..', '..')

const REQUIRED_SUBSTRINGS = new Map([
  [path.join(ROOT, 'SKILL.md'), [
    'one official skill',
    '`a2_cli`',
    'friend_skills/',
    '`a2_cli friend msg`',
    '`a2_cli learning start`',
  ]],
  [path.join(ROOT, 'package.json'), [
    '"name": "agentsquared-official-skills"',
    '"version": "1.0.0"',
    '"a2_cli": "./scripts/a2_cli.mjs"',
  ]],
  [path.join(ROOT, 'scripts', 'a2_cli.mjs'), [
    'runA2Cli',
  ]],
  [path.join(ROOT, 'scripts', 'lib', 'a2_cli_core.mjs'), [
    'friend msg',
    'message send',
    'learning start',
    'skill-file',
    'loadSharedSkillFile',
  ]],
  [path.join(ROOT, 'references', 'a2_cli.md'), [
    '`a2_cli`',
    'single official deterministic command surface',
    'a2_cli friend msg',
    '--skill-file friend_skills/<skill-name>/skill.md',
  ]],
  [path.join(ROOT, 'references', 'friend_skill_library.md'), [
    '`friend_skills/`',
    'lowercase `skill.md`',
    '`a2_cli friend msg`',
  ]],
  [path.join(ROOT, 'catalog', 'index.json'), [
    '"version": "1.0.0"',
    '"agentsquared-official-skills": "SKILL.md"',
  ]],
  [path.join(ROOT, 'runtime', 'guide.md'), [
    'relay MCP',
    'shared gateway lifecycle',
    'trusted peer-session reuse',
  ]],
  [path.join(ROOT, 'init', 'guide.md'), [
    'Initialize or re-initialize the local AgentSquared runtime',
    'GET /health',
    'GET /inbox/index',
  ]],
  [path.join(ROOT, 'policy', 'guide.md'), [
    'Human-rooted trust model',
    'public-safe projections',
    'relay for coordination only',
  ]],
  [path.join(ROOT, 'onboarding', 'guide.md'), [
    'registration',
    'runtime init',
  ]],
  [path.join(ROOT, 'friend_skills', 'friend-im', 'skill.md'), [
    'friend-im',
    'private peer payload',
  ]],
  [path.join(ROOT, 'friend_skills', 'agent-mutual-learning', 'skill.md'), [
    'mutual-learning',
    'private session',
  ]],
  [path.join(ROOT, 'friend_skills', 'friend-im', 'skill.md'), [
    'name: friend-im',
  ]],
  [path.join(ROOT, 'friend_skills', 'agent-mutual-learning', 'skill.md'), [
    'name: agent-mutual-learning',
  ]],
  [path.join(ROOT, 'runtime', 'scripts', 'serve_gateway.mjs'), [
    '/health',
    '/inbox/index',
    '/connect',
    'runtimeState',
  ]],
  [path.join(ROOT, 'runtime', 'adapters', 'openclaw', 'adapter.mjs'), [
    'agent.wait',
    'chat.history',
    'sharedSkillDocument',
  ]],
])

const FORBIDDEN_SKILL_PATHS = [
  path.join(ROOT, 'Base', 'SKILL.md'),
  path.join(ROOT, 'Friends', 'SKILL.md'),
  path.join(ROOT, 'Identity', 'SKILL.md'),
  path.join(ROOT, 'Maintainers', 'agentsquared-skills-helper', 'SKILL.md'),
]

const FORBIDDEN_DIRECTORIES = [
  path.join(ROOT, 'Base'),
  path.join(ROOT, 'Friends'),
  path.join(ROOT, 'Identity'),
]

function main() {
  const errors = []

  for (const [filePath, snippets] of REQUIRED_SUBSTRINGS.entries()) {
    if (!fs.existsSync(filePath)) {
      errors.push(`${filePath}: missing required file`)
      continue
    }
    const text = fs.readFileSync(filePath, 'utf8')
    for (const snippet of snippets) {
      if (!text.includes(snippet)) {
        errors.push(`${filePath}: missing required text: ${snippet}`)
      }
    }
  }

  for (const filePath of FORBIDDEN_SKILL_PATHS) {
    if (fs.existsSync(filePath)) {
      errors.push(`${filePath}: unexpected nested SKILL.md; the repository should expose only the root SKILL.md`)
    }
  }

  for (const dirPath of FORBIDDEN_DIRECTORIES) {
    if (fs.existsSync(dirPath)) {
      errors.push(`${dirPath}: unexpected legacy top-level directory; migrate content into the flat root-level structure`)
    }
  }

  const rootSkills = []
  for (const entry of fs.readdirSync(ROOT, { withFileTypes: true })) {
    if (entry.isFile() && entry.name === 'SKILL.md') {
      rootSkills.push(path.join(ROOT, entry.name))
    }
  }
  if (rootSkills.length !== 1) {
    errors.push(`expected exactly one root SKILL.md, found ${rootSkills.length}`)
  }

  if (errors.length > 0) {
    console.error('AgentSquared Skills runtime-contract validation failed:')
    for (const error of errors) {
      console.error(`- ${error}`)
    }
    process.exit(1)
  }

  console.log('AgentSquared Skills runtime-contract validation passed.')
}

main()
